Posts/2021/10/parsing-netstat.md

```diff
@@ -0,0 +1,207 @@
+lines of text. We know from our analysis that the table header is on the fourth line (third line for
+Linux), so the data starts on the next line.
+
+```powershell
+PS> $lines = netstat -a
+PS> $lines.count
+440
+PS> $lines[3]
+  Proto  Local Address          Foreign Address        State
+PS> $lines[4]
+  TCP    0.0.0.0:80             0.0.0.0:0              LISTENING
+```
+
+To parse the rows into the individual columns of data we need to use the space character as a
+delimiter to split the line. Since the number of spaces between columns is variable, the split
+operation creates empty fields between the data. We can filter those empty fields out with a
+`Where-Object`. For example:
+
+```powershell
+$columns = ($lines[4] -split ' ').Trim() | Where-Object {$_ }
+$columns
+TCP
+0.0.0.0:80
+0.0.0.0:0
+LISTENING
+```
+
+In this example, the `Trim()` method trims off leading and trailing spaces. This ensures that the
+fields between the columns become empty strings.
+
+## Step 4 - Output the object
+
+The only thing left to do now is to create a PowerShell object that contains the parsed data. Let's
+put this all together.
+
+```powershell
+function parseNetstat {
+    param([object[]]$Lines)
+
+    if ($IsWindows) {
+        $skip = 4
+    } else {
+        $skip = 3
+    }
+
+    $Lines | Select-Object -Skip $skip | ForEach-Object {
+        $columns = ($_ -split ' ').Trim() | Where-Object {$_ }
+        [pscustomobject]@{
+            Protocol = $columns[0]
+            LocalAddress = $columns[1]
+            RemoteAddress = $columns[2]
+            State = $columns[3]
+        }
+    }
+}
+
+parseNetstat (netstat) | Select-Object -Last 10
+```
+
+For this example, I limit the output to the last 10 rows.
+
+```powershell-console
+Protocol LocalAddress                                   RemoteAddress                 State
+-------- ------------                                   -------------                 -----
+TCP      [2600:6c56:7e00:78d:e1e8:756c:d2be:42da]:61001 [2603:1036:303:3000::2]:https TIME_WAIT
+TCP      [2600:6c56:7e00:78d:e1e8:756c:d2be:42da]:61018 [2603:1030:408::401]:https    ESTABLISHED
+TCP      [2600:6c56:7e00:78d:e1e8:756c:d2be:42da]:61293 [2603:1036:303:3000::2]:https ESTABLISHED
+TCP      [2600:6c56:7e00:78d:e1e8:756c:d2be:42da]:62640 [2603:1036:303:3c33::2]:https ESTABLISHED
+TCP      [2600:6c56:7e00:78d:e1e8:756c:d2be:42da]:62643 [2603:1036:303:3c04::2]:https ESTABLISHED
+TCP      [2600:6c56:7e00:78d:e1e8:756c:d2be:42da]:62659 [2603:1036:303:3050::2]:https TIME_WAIT
+TCP      [2600:6c56:7e00:78d:e1e8:756c:d2be:42da]:64886 ord37s36-in-x0d:https         ESTABLISHED
+TCP      [2600:6c56:7e00:78d:e1e8:756c:d2be:42da]:64887 [2603:1036:404:8e::2]:https   TIME_WAIT
+TCP      [2600:6c56:7e00:78d:e1e8:756c:d2be:42da]:64901 [2620:1ec:21::18]:https       ESTABLISHED
+TCP      [2600:6c56:7e00:78d:e1e8:756c:d2be:42da]:65492 ord30s21-in-x0e:https         TIME_WAIT
+```
+
```

Comment:
At this point - I would add a line that confirms success with the reader before moving to the advanced part where you further clean it up. -- "Success! If you're working with Crescendo, now you can move to creating the configuration file."

Reply:
Added a note